File: src/common/py/core/logging/logger_proxy.py
import typing

from .logger import Logger


class LoggerProxy:
    """
    A proxy to automatically pass extra parameters to a logger.
    
    This class allows us to store additional, fixed parameters passed to an existing logger, avoiding the need to use
    a new logger instance. It offers the same public interface as an actual ``Logger`` and can thus be used like a
    *real* logger.
    
    Args:
        logger: The logger to use.
    """
    def __init__(self, logger: Logger):
        """
        Args:
            logger: The logger to use.
        """
        self._logger = logger
        
        self._auto_params: typing.Dict[str, typing.Any] = {}
        
    def add_param(self, name: str, value: typing.Any) -> None:
        """
        Adds a new paramter that is always automatically passed to the logger.
        
        Args:
            name: The name of the parameter.
            value: Its value.
        """
        self._auto_params[name] = value
        
    def remove_param(self, name: str) -> None:
        """
        Removes a parameter that has been added previously.
        """
        try:
            self._auto_params.pop(name)
        except KeyError:
            pass
        
    def clear_params(self) -> None:
        """
        Removes all stored parameters.
        """
        self._auto_params = {}
    
    def debug(self, msg: str, *, scope: str | None = None, **kwargs) -> None:
        """
        Logs a debugging message.
        
        Args:
            msg: The text to log.
            scope: The scope of the entry.
            **kwargs: Any additional parameters.
        """
        self._logger.debug(msg, scope=scope, **(kwargs | self._auto_params))
    
    def info(self, msg: str, *, scope: str | None = None, **kwargs) -> None:
        """
        Logs an information message.

        Args:
            msg: The text to log.
            scope: The scope of the entry.
            **kwargs: Any additional parameters.
        """
        self._logger.info(msg, scope=scope, **(kwargs | self._auto_params))
    
    def warning(self, msg: str, *, scope: str | None = None, **kwargs) -> None:
        """
        Logs a warning message.

        Args:
            msg: The text to log.
            scope: The scope of the entry.
            **kwargs: Any additional parameters.
        """
        self._logger.warning(msg, scope=scope, **(kwargs | self._auto_params))
    
    def error(self, msg: str, *, scope: str | None = None, **kwargs) -> None:
        """
        Logs an error message.

        Args:
            msg: The text to log.
            scope: The scope of the entry.
            **kwargs: Any additional parameters.
        """
        self._logger.error(msg, scope=scope, **(kwargs | self._auto_params))

File: src/common/web/core/logging/LoggerProxy.ts
import { Logger } from "./Logger";

/**
 * A proxy to automatically pass extra parameters to a logger.
 *
 * This class allows us to store additional, fixed parameters passed to an existing logger, avoiding the need to use
 * a new logger instance. It offers the same public interface as an actual ``Logger`` and can thus be used like a
 * *real* logger.
 */
export class LoggerProxy {
    private readonly _logger: Logger;

    private _autoParams: Record<string, any> = {};

    /**
     * @param logger - The logger to use.
     */
    public constructor(logger: Logger) {
        this._logger = logger;
    }

    /**
     * Adds a new paramter that is always automatically passed to the logger.
     *
     * @param name - The name of the parameter.
     * @param value - Its value.
     */
    public addParam(name: string, value: any): void {
        this._autoParams[name] = value;
    }

    /**
     * Removes a parameter that has been added previously.
     *
     * @param name - The name of the parameter.
     */
    public removeParam(name: string): void {
        delete this._autoParams[name];
    }

    /**
     * Removes all stored parameters.
     */
    public clearParams(): void {
        this._autoParams = {};
    }

    /**
     * Logs a debugging message.
     *
     * @param msg - The text to log.
     * @param scope - The scope of the entry.
     * @param params - Any additional parameters.
     */
    public debug(msg: string, scope: string = "", params: Record<string, any> = {}): void {
        this._logger.debug(msg, scope, { ...this._autoParams, ...params });
    }

    /**
     * Logs an info message.
     *
     * @param msg - The text to log.
     * @param scope - The scope of the entry.
     * @param params - Any additional parameters.
     */
    public info(msg: string, scope: string = "", params: Record<string, any> = {}): void {
        this._logger.info(msg, scope, { ...this._autoParams, ...params });
    }

    /**
     * Logs a warning message.
     *
     * @param msg - The text to log.
     * @param scope - The scope of the entry.
     * @param params - Any additional parameters.
     */
    public warning(msg: string, scope: string = "", params: Record<string, any> = {}): void {
        this._logger.warning(msg, scope, { ...this._autoParams, ...params });
    }

    /**
     * Logs an error message.
     *
     * @param msg - The text to log.
     * @param scope - The scope of the entry.
     * @param params - Any additional parameters.
     */
    public error(msg: string, scope: string = "", params: Record<string, any> = {}): void {
        this._logger.error(msg, scope, { ...this._autoParams, ...params });
    }
}
